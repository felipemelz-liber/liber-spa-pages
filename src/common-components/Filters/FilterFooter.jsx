import React from 'react';
import PropTypes from 'prop-types';
import { Footer, FilterButton, ClearButton } from './Filters.styles';

const FilterFooter = ({ disabled, onFilter, onClear }) => {
  return (
    <Footer>
      <FilterButton disabled={disabled} onClick={onFilter}>
        FILTRAR
      </FilterButton>
      <ClearButton onClick={onClear} disabled={disabled}>
        LIMPAR
      </ClearButton>
    </Footer>
  );
};

FilterFooter.propTypes = {
  disabled: PropTypes.bool,
  onFilter: PropTypes.func,
  onClear: PropTypes.func,
};

FilterFooter.defaultProps = {
  disabled: false,
  onFilter: () => null,
  onClear: () => null,
};

export default FilterFooter;
