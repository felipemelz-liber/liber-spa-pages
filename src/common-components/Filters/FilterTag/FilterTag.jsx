import React from 'react';
import PropTypes from 'prop-types';
import {
  FilterTag as Tag,
  SelectedLabelContainer,
  SelectedLabelOption,
  SelectedSwitch,
} from './FilterTag.styles';

const isSelected = list => {
  if (list) {
    return list.filter(item => Boolean(item)).length !== 0;
  }
  return false;
};

const format = list => {
  return list.join(' ');
};

const FilterTag = ({ selected, formatSelected, label, onClear, show, ...props }) => {
  const [hover, setHover] = React.useState(false);
  let tagProps;
  if (isSelected(selected)) {
    const nodeLabel = (
      <SelectedLabelContainer>
        <SelectedSwitch switch={hover || show}>
          <SelectedLabelOption>{formatSelected(selected)}</SelectedLabelOption>
          <SelectedLabelOption>{label}</SelectedLabelOption>
        </SelectedSwitch>
      </SelectedLabelContainer>
    );
    tagProps = {
      label: nodeLabel,
      closable: true,
      onClose: onClear,
      onMouseEnter: () => setHover(true),
      onMouseLeave: () => setHover(false),
      gray: !show,
      show,
      ...props,
    };
  } else {
    tagProps = { label, show, ...props };
  }
  return <Tag {...tagProps} />;
};

FilterTag.propTypes = {
  selected: PropTypes.arrayOf(PropTypes.any),
  formatSelected: PropTypes.func,
  onClear: PropTypes.func,
  label: PropTypes.string,
  show: PropTypes.bool,
};

FilterTag.defaultProps = {
  selected: [],
  formatSelected: format,
  onClear: () => {},
  show: false,
  label: '',
};

export default FilterTag;
