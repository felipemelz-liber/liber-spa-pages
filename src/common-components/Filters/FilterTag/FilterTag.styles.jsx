import styled, { css } from 'styled-components';
import { ModalTag } from 'liber-components';

const isGray = ({ gray }) => {
  if (gray)
    return css`
      background-color: #f4f5f7;
      border-color: #f4f5f7;
    `;
  return null;
};

export const FilterTag = styled(ModalTag)`
  font-size: 16px;
  min-height: 36px;
  border-radius: 4px;
  font-weight: 400;
  ${isGray}
`;

export const SelectedLabelContainer = styled.div`
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
`;

export const SelectedSwitch = styled.div`
  height: 52px;
  width: 100%;
  padding-top: 6px;
  display: flex;
  flex-direction: column;
  transform: ${props => (props.switch ? 'translateY(-26px)' : 'translateY(0px)')};
  transition: transform 250ms ease-in-out;
`;

export const SelectedLabelOption = styled.span`
  width: 100%;
  height: 26px;
  display: flex;
  justify-content: center;
  align-items: center;
`;
