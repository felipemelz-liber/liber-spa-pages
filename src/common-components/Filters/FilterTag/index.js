export { default } from './FilterTag.jsx';
