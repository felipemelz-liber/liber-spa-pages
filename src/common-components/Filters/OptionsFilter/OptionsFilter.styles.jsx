import styled from 'styled-components';

export const OptionsContainer = styled.div`
  width: 320px;
  display: flex;
  flex-wrap: wrap;
  padding: 20px;
  padding-bottom: 8px;

  & > div {
    margin-bottom: 12px;
  }
`;
