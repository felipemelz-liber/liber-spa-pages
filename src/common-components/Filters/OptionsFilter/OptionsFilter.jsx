import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import FilterTag from '../FilterTag';
import SelectTag from '../../SelectTag/SelectTag';
import FilterFooter from '../FilterFooter';
import { OptionsContainer } from './OptionsFilter.styles';

const handleFilter = (option, selectedState) => () => {
  const [selectedOptions, setSelectedOptions] = selectedState;
  let finalSelection = [];

  if (selectedOptions.includes(option)) {
    finalSelection = selectedOptions.filter(selectedOption => selectedOption !== option);
  } else {
    finalSelection = [...selectedOptions, option];
  }

  setSelectedOptions(finalSelection);
};

const formatLabel = label => list => {
  const [selectedOptions] = list;

  return `${label} (${selectedOptions.length})`;
};

const OptionsFilter = ({ options, onFilter, onClear, filterValues, label }) => {
  const [open, setOpen] = useState(false);
  const selectedState = useState(filterValues || []);
  const [selectedOptions, setSelectedOptions] = selectedState;

  useEffect(() => {
    if (!filterValues) {
      setSelectedOptions([]);
    }
  }, [filterValues]);

  return (
    <FilterTag
      selected={[filterValues]}
      onClear={onClear}
      show={open}
      label={label}
      formatSelected={formatLabel(label)}
      onChange={setOpen}
    >
      <OptionsContainer>
        {options.map(({ value, option }) => (
          <SelectTag
            key={`option-${value}`}
            onChange={handleFilter(value, selectedState)}
            selected={selectedOptions.includes(value)}
          >
            {option}
          </SelectTag>
        ))}
      </OptionsContainer>
      <FilterFooter
        disabled={selectedOptions.length === 0}
        onFilter={() => {
          setOpen(false);
          onFilter(selectedOptions);
        }}
        onClear={onClear}
      />
    </FilterTag>
  );
};

OptionsFilter.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({ value: PropTypes.any, option: PropTypes.string })),
  filterValues: PropTypes.arrayOf(PropTypes.any),
  onFilter: PropTypes.func,
  onClear: PropTypes.func,
  label: PropTypes.string,
};

OptionsFilter.defaultProps = {
  options: [],
  filterValues: null,
  onFilter: () => {},
  onClear: () => {},
  label: '',
};

export default OptionsFilter;
