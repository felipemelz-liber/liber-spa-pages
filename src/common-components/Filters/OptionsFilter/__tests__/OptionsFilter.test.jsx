import React from 'react';
import sinon from 'sinon';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import OptionsFilter from '../OptionsFilter';
import FilterTag from '../../FilterTag/FilterTag';
import FilterFooter from '../../FilterFooter';

describe('OptionsFilter tests', () => {
  configure({ adapter: new Adapter() });

  const options = [{ option: 'ConcluÃ­da', value: 1 }, { option: 'Cancelada', value: 2 }];
  const onClear = sinon.spy();
  const onFilter = sinon.spy();
  const props = {
    filterValues: [1],
    options,
    onClear,
    onFilter,
    label: 'mock',
  };

  afterEach(() => {
    sinon.resetHistory();
  });

  it('should render correctly without props', () => {
    const wrapper = shallow(<OptionsFilter />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly with props', () => {
    const wrapper = shallow(<OptionsFilter {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call on clear on filter tag clear', () => {
    const wrapper = shallow(<OptionsFilter {...props} />);
    const filterTag = wrapper.find(FilterTag);

    filterTag.simulate('clear');

    expect(onClear.called).toBeTruthy();
  });

  it('should call on clear on filter footer clear', () => {
    const wrapper = shallow(<OptionsFilter {...props} />);
    const filterFooter = wrapper.find(FilterFooter);

    filterFooter.simulate('clear');

    expect(onClear.called).toBeTruthy();
  });

  it('should call onFilter on filter footer filter', () => {
    const wrapper = shallow(<OptionsFilter {...props} />);
    const footer = wrapper.find(FilterFooter);

    footer.simulate('filter');

    expect(onFilter.called).toBeTruthy();
  });
});
