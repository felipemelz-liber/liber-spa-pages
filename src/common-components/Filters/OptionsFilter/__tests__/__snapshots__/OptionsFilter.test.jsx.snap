// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsFilter tests should render correctly with props 1`] = `
<FilterTag
  formatSelected={[Function]}
  label="mock"
  onChange={[Function]}
  onClear={[Function]}
  selected={
    Array [
      Array [
        1,
      ],
    ]
  }
  show={false}
>
  <styled.div>
    <SelectTag
      key="option-1"
      onChange={[Function]}
      selected={true}
    >
      Conclu√≠da
    </SelectTag>
    <SelectTag
      key="option-2"
      onChange={[Function]}
      selected={false}
    >
      Cancelada
    </SelectTag>
  </styled.div>
  <FilterFooter
    disabled={false}
    onClear={[Function]}
    onFilter={[Function]}
  />
</FilterTag>
`;

exports[`OptionsFilter tests should render correctly without props 1`] = `
<FilterTag
  formatSelected={[Function]}
  label=""
  onChange={[Function]}
  onClear={[Function]}
  selected={
    Array [
      null,
    ]
  }
  show={false}
>
  <styled.div />
  <FilterFooter
    disabled={true}
    onClear={[Function]}
    onFilter={[Function]}
  />
</FilterTag>
`;
